site/source/docs/api_reference/wasm_workers.rst

```diff
@@ -0,0 +1,355 @@
+    emscripten_wasm_worker_post_function_v(worker, run_in_worker);
+  }
+
+Build the code by passing the Emscripten flag ``-sWASM_WORKERS=1`` at both compile
+and link steps. The example code creates a new Worker on the main browser thread,
+which shares the same WebAssembly.Module and WebAssembly.Memory object. Then a
+``postMessage()`` is passed to the Worker to ask it to execute the function
+``run_in_worker()`` to print a string.
+
+For more complex threading scenarios, see the functions ``emscripten_create_wasm_worker_with_tls()``
+and ``emscripten_create_wasm_worker_no_tls()`` on how to manage the memory resources that
+the thread will use.
+
+Introduction
+============
+
+In WebAssembly programs, the Memory object that contains the application state can be
+shared across multiple Workers. This enables direct, high performance (and if explicit
+care is not taken, racy!) access to synchronously share data state between multiple
+Workers (shared state multithreading).
+
+Emscripten supports two multithreading APIs to leverage this web feature:
+ - POSIX Threads (Pthreads) API, and
+ - Wasm Workers API.
+
+The Pthreads API has a long history with native C programming and the POSIX standard,
+while Wasm Workers API is unique to Emscripten compiler only.
+
+These two APIs provide largely the same feature set, but have important differences,
+which this documentation seeks to explain to help decide which API one should target.
+
+Pthreads vs Wasm Workers: Which One to Use?
+===========================================
+
+The intended audience and use cases of these two multithreading APIs are slightly
+different.
+
+The focus on Pthreads API is on portability and cross-platform compatibility. This API
+is best used in scenarios where portability is most important, e.g. when a codebase is
+cross-compiled to multiple platforms, like building both a native Linux x64 executable and an
+Emscripten WebAssembly based web site.
+
+Pthreads API in Emscripten seeks to carefully emulate compatibility and the features that
+the native Pthreads platforms already provide. This helps porting large C/C++ codebases
+over to WebAssembly.
+
+Wasm Workers API on the other hand seeks to provide a "direct mapping" to the web
+multithreading primitives as they exist on the web, and call it a day. If an application
+is only developed to target WebAssembly, and portability is not a concern, then using Wasm
+Workers can provide great benefits in the form of simpler compiled output, less complexity,
+smaller code size and possibly better performance.
+
+However this benefit might not be an obvious win. The Pthreads API was designed to be useful
+from the synchronous C/C++ language, whereas Web Workers are designed to be useful from
+asynchronous JavaScript. WebAssembly C/C++ programs can find themselves somewhere in
+the middle.
+
+Pthreads and Wasm Workers share several similarities:
+
+ * Both can use emscripten_atomic_* Atomics API,
```

Comment:
We do want to support both. C++11 standard atomics contain parameters that are not supported by WebAssembly, so it is good to have an explicit API that matches exactly the operations exposed by Wasm. I see I did not list C11 and C++11 Atomics, added those to the list now.

Reply:
> it is good to have an explicit API that matches exactly the operations exposed by Wasm.

Why? Even if Wasm doesn't support memory orderings besides seqcst yet, using them at the source level better captures programmer intent and has a better forward-compatibility story for when Wasm does support more memory orderings. This is perhaps a separate discussion, but it would be better IMO to soft-deprecate non-portable Emscripten APIs that have portable alternatives.

Reply:
> using them at the source level better captures programmer intent

This is a dual edged sword. I don't believe that is necessarily the case. I would argue that code targeting explicit wasm atomics better captures the programmer intent, since when one reads the code authored against those, it is clear to read that the programmer has been thinking about targeting the web with those operations specifically.

Reading code that uses other intrinsics APIs, there is no guarantee that code has ever been tested to run on the web.

> and has a better forward-compatibility story for when Wasm does support more memory orderings

This also has its pros and cons I think. It can be argued that the explicit atomics have better forward compatibility:

If wasm gains new memory orderings, those cannot break the existing Atomics functions, the existing operations will still be guaranteed to be preserved.

As a developer, if I write against those operations specifically, I have a guarantee that several years into the future, that code will still build exactly like the day that I authored it, and will run exactly like I tested it to do.

If I write a relaxed atomic operation today with the C++11 intrinsics, then test it on the web, observe that it passes testing, and then ship it; and several years later I have left the company, and wasm gains relaxed intrinsics. All of a sudden recompiling the code will cause those relaxed memory orderings to be now in effect. If I had a subtle bug in my reasoning, it would have not gotten caught back when I authored it. (even more tricky with atomics is that people might have tested it against native x64, but then it could fail only on ARM browsers given the different semantics that x64 vs ARM have)

Also, people have to consider the issues of shipping code that they build to their users, and which browser versions they target. If I author code using explicit Emscripten specific intrinsics, the code will compile the same in the future even when new relaxed operations become available, and the code will still target the same set of browser versions exactly as it used to, without removing backwards compatibility.

So by targeting the explicit Emscripten atomic operations I have an explicit say and a guarantee how the code will compile like in the future.

I do appreciate the importance also of being able to write portable code that will compile across all platforms. I am not at all saying that using other atomics is a bad thing, sometimes those are greatly preferable instead (in fact, we use both in different parts of the codebase), but I don't think it is the only solution to have; both APIs are definitely warranted in my opinion.